provider/oracle/images.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package oracle
+
+import (
+	"strings"
+
+	oci "github.com/hoenirvili/go-oracle-cloud/api"
+	"github.com/juju/errors"
+	"github.com/juju/juju/constraints"
+	"github.com/juju/juju/environs/imagemetadata"
+	"github.com/juju/juju/environs/instances"
+)
+
+// getInstanceTypes retrives all types that the oracle provider supports
+func getInstanceTypes(c *oci.Client) ([]instances.InstanceType, error) {
+	if c == nil {
+		return nil, errors.Errorf("cannot use nil client")
+	}
+
+	// make api request to oracle cloud to give us a list of
+	// all supported shapes
+	shapes, err := c.AllShapeDetails()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// so let's transform these into juju complaint
+	// instance types
+	onlyArch := []string{"amd64"}
+
+	types := make([]instances.InstanceType, len(shapes.Result), len(shapes.Result))
+	for key, val := range shapes.Result {
+		types[key].Name = val.Name
+		types[key].Arches = onlyArch
+		types[key].Mem = val.Ram
+		types[key].CpuCores = uint64(val.Cpus)
+		types[key].RootDisk = val.Root_disk_size
+	}
+
+	return types, nil
+}
+
+// findInstanceSpec returns an *InstanceType imagelist name satisfying the supplied instanceConstraint
+func findInstanceSpec(
+	c *oci.Client,
+	allImageMetadata []*imagemetadata.ImageMetadata,
+	instanceType []instances.InstanceType,
+	is *instances.InstanceConstraint,
+) (*instances.InstanceSpec, string, error) {
+
+	logger.Debugf("recived %d image(s)", len(allImageMetadata))
+
+	images := instances.ImageMetadataToImages(allImageMetadata)
+	spec, err := instances.FindInstanceSpec(images, is, instanceType)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+
+	imagelist, err := getImageName(c, spec.Image.Id)
+	if err != nil {
+		return nil, "", errors.Trace(err)
+	}
+
+	return spec, imagelist, nil
+}
+
+// checkImageList creates a list of metadata images from the oracle image list
+func checkImageList(c *oci.Client, cons constraints.Value) ([]*imagemetadata.ImageMetadata, error) {
+
+	// take a list of all images that are in the oracle cloud account
+	resp, err := c.AllImageList()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	images := make([]*imagemetadata.ImageMetadata, 0, len(resp.Result))
+
+	for _, val := range resp.Result {
```

Comment:
there's a lot of magic numbers here, can you please add comments on the expected results format. pointers to docs?